execute/profiler.go

```diff
@@ -79,25 +83,66 @@ func (t *OperatorProfilingSpan) FinishWithOptions(opts opentracing.FinishOptions
 
 const OperatorProfilerContextKey = "operator-profiler"
 
+type operatorProfilingResultAggregate struct {
+	operationType string
+	label         string
+	resultCount   int64
+	resultMin     int64
+	resultMax     int64
+	resultSum     int64
+	resultMean    float64
+}
+
+type operatorProfilerGroupedAggregates = map[string]map[string]operatorProfilingResultAggregate
+
 type OperatorProfiler struct {
-	results []OperatorProfilingResult
 	// Receive the profiling results from the spans.
-	ch chan OperatorProfilingResult
-	mu sync.Mutex
+	chIn  chan OperatorProfilingResult
+	chOut chan operatorProfilingResultAggregate
 }
 
 func createOperatorProfiler() Profiler {
 	p := &OperatorProfiler{
-		results: make([]OperatorProfilingResult, 0),
-		ch:      make(chan OperatorProfilingResult),
+		chIn:  make(chan OperatorProfilingResult),
+		chOut: make(chan operatorProfilingResultAggregate),
 	}
 	go func(p *OperatorProfiler) {
-		for result := range p.ch {
-			p.mu.Lock()
-			p.results = append(p.results, result)
-			p.mu.Unlock()
+		aggs := make(operatorProfilerGroupedAggregates)
+		for result := range p.chIn {
+			// Because we're using nested maps, we need to check if the inner map
+			// exists before we started manipulating it. If it doesn't, we create one.
+			_, ok := aggs[result.Type]
+			if !ok {
+				aggs[result.Type] = make(map[string]operatorProfilingResultAggregate)
+			}
+			a := aggs[result.Type][result.Label]
+
+			// Aggregate the results
+			a.resultCount += 1
+			duration := result.Stop.Sub(result.Start).Nanoseconds()
+			if duration > a.resultMax {
+				a.resultMax = duration
+			}
+			if duration < a.resultMin || a.resultMin == 0 {
+				a.resultMin = duration
+			}
+			a.resultSum += duration
+			aggs[result.Type][result.Label] = a
```

Comment:
You can let the map store the pointer: `map[string]*operatorProfilingResultAggregate`, so you can directly mutate the data in `a` and don't need to assign it again to the map `aggs[result.Type][result.Label] = a`